Add insert

mpospelov opened the pull request.

mpospelov committed:
implement insert_after

mpospelov committed:
implement insert before

mpospelov committed:
test block execution count

mpospelov committed:
test block execution count

mpospelov committed:
add link to list + add methods for adding nodes via single node

mpospelov committed:
Merge branch 'add_delete' into add_insert

mpospelov committed:
Merge branch 'adapt_api' into add_insert

spectator commented:
Hey @mpospelov ! appreciate your work on this, but I'm running into a few issues. Sorry, I should have done a better job reviewing previous pull request, so I have to pile up everything in here.

I have troubles using `delete` and `delete_all` methods outside of tests on current master, here is an example

```
~/P/s/linked-list git:master ❯❯❯ console
irb(main):001:0> list = LinkedList::List.new
=> #<LinkedList::List:0x3fd282088d6c []>
irb(main):002:0> list.push('foo')
=> #<LinkedList::List:0x3fd282088d6c ["foo"]>
irb(main):003:0> list.push('foo')
=> #<LinkedList::List:0x3fd282088d6c ["foo", "foo"]>
irb(main):004:0> list.push('bar')
=> #<LinkedList::List:0x3fd282088d6c ["foo", "foo", "bar"]>
irb(main):005:0> list.delete('foo')
Traceback (most recent call last):
        5: from ./bin/console:8:in `<main>'
        4: from (irb):5
        3: from /linked-list/lib/linked-list/list.rb:78:in `delete'
        2: from /linked-list/lib/linked-list/list.rb:78:in `tap'
        1: from /linked-list/lib/linked-list/list.rb:79:in `block in delete'
NoMethodError (undefined method `blank?' for #<LinkedList::Node:0x00007fa5051adf68>)
```

This pull request obviously have the same issues for new methods -- `insert_before` and `insert_after`.

`blank?` method exists only in environments with `ActiveSupport` present, there is no such method in Ruby.

For this pull request it looks like that `Node` and `List` responsibilities are mixed up in the `Node` now. `Node` is now not only "knows" about the `List` but keeps the reference to the list. I prefer that `Node` stays as is and `List` does all node manipulations.

mpospelov commented:
ok sure no problem let's change this 😉

mpospelov committed:
Merge remote-tracking branch 'upstream/master' into add_insert

mpospelov committed:
fix blank issues + remove list responsibility from Node class

mpospelov commented:
@spectator I introduced new methods `insert_after_node`, `instert_before_node` as I need them in conjunction with `each_node`, so to not iterate each time when I want to insert data in the list.

I will add tests and doc for them latter today but you can check the code 😉 

Also, I removed all `#blank?` calls, it was strange why tests didn't fail then, I don't see `ActiveSupport` in dependancies.

spectator commented:
@mpospelov `delete` and `delete_all` work now 👍 but `insert_after` and `insert_before` don't, both simply return `nil`

mpospelov commented:
@spectator I found that this `#blank?` and other `ActiveSupport` like behaviour comes from `tins` which is `coveralls` dependency.
So I added `CI` env variable to Travis and also add a condition for requiring `coveralls`.
Can you please give a code example of `insert_after`/`insert_before` issue?

mpospelov committed:
fix requiring coveralls localy

spectator commented:
Good idea to isolate Coveralls!

Here is what tried for `after_insert`

```ruby
irb(main):009:0> list = LinkedList::List.new
=> #<LinkedList::List:0x3fbfef4fefcc []>
irb(main):010:0> list.push(1)
=> #<LinkedList::List:0x3fbfef4fefcc [1]>
irb(main):011:0> list.push(2)
=> #<LinkedList::List:0x3fbfef4fefcc [1, 2]>
irb(main):012:0> list.insert_after(1, 3)
=> nil
irb(main):013:0> list
=> #<LinkedList::List:0x3fbfef4fefcc [1, 2]>
```

mpospelov commented:
The example you send is correct, you are trying to insert `1` after `3` and as there is no `3` in the list you got `nil`. You can pass the condition to insert after as a block or as an exact value
```
list.insert_after(1) { |d| d == 3 }
```
and to make possible to pass block or value I made value last argument(it's `nil` by default and I use `__to_matcher` helper to convert an exact value to block). 
If you have better suggestions for method semantic I would like to know them 😉 
maybe it will be more readable and developer friendly?
```
insert(3, after: 1)
insert(3, after: (d) -> d == 1)
insert(3, before: 1)
insert(3, before: (d) -> d == 1)
```

spectator commented:
🤦‍♂️ I'm sorry, for some reason I thought first argument is "index", maybe because that is how `Array#insert` works. Maybe keep order of arguments similar to that?

spectator commented on lib/linked-list/list.rb:
```diff
@@ -191,19 +213,39 @@ def to_list
     end
 
-    private
+    def insert_after_node(data, node)
```
These three methods `insert_after_node`, `insert_before_node` and `unlink_node` are better to be private I think. Also, let's keep method names similar to others -- `__insert_after`, `__insert_before` and `__unlink`.

mpospelov replied:
I need them in conjunction with `#each_node`, as otherwise, I need to go throw list each time, which increase the complexity

spectator replied:
I'm not sure I understand you meant by this. Would elaborate please?

mpospelov replied:
So my use case is to iterate over nodes and by some decision insert data before or after that nodes, like this:
```
list.each_node do |node|
  my_data = ...
  if something?
     list.insert_after_node(node, my_data)
  else
     list.insert_before_node(node, my_data)
  end
end
```

if I do some like this
```
list.each_node do |node|
  my_data = ...
  if something?
     list.insert(my_data, after: ->(d) { node.data == d })
  else
     list.insert(my_data, before: ->(d) { node.data == d })
  end
end
```
it will invoke iteration over list each time and instead of `O(n)` complexity in first case I will get `O(n**2)`

spectator replied:
Hmm, I only meant to make those methods private and change names to stay consistent with existing naming. That should not affect how you iterate over a list, right?

mpospelov replied:
If make them private I should do `list.send(: __insert_after, node, data)` which looks weird, I just need a public API to insert before/after node during `each_node`

spectator replied:
But where would you do that? There used to be something like that in `Node`, but now it is back to `List` which doesn't need explicit caller. Am I missing something?

mpospelov replied:
I send you an example of expected from my side usage of linked-list api
```
list.each_node do |node|
  my_data = ...
  if something?
     list.insert_after_node(node, my_data)
  else
     list.insert_before_node(node, my_data)
  end
end
```

spectator replied:
I see now, so you need both `insert('foo', after: 'bar')` and `insert_after_node(data, node)`. Sure, let's keep them public then, but unlink doesn't have to be. I just don't want to increase public methods surface unnecessarily.

BTW, if it something you are willing to share, what do you use linked list for?

mpospelov replied:
I have a schedule data coming from database like: 
```
guy1 busy: [
  { from: '10:30', to: '12:00', reason: 'dinner' }, 
  { from: '9:00', to: '11:00', reason: 'works at object1' }, 
  { from: '0:00', to: '08:00', reason: 'working hours'}, 
  ...
]
``` 

My task is to find all overlapping data points and make them follow one by one so there won't be overlaps, also this data has types which should be merged too

So there are some cases when one data point fully covered by another:
```
slot1: 09:00-13:00, is_home: true, is_in_salon: false 
slot2: 11:00 - 12:00, is_home: true, is_in_salon: true
```

In that case I need to produce 3 slots
```
new_slot1: 09:00-11:00, is_home: true, is_in_salon: false 
new_slot2: 11:00 - 12:00, is_home: true, is_in_salon: true
new_slot3: 12:00 - 13:00, is_home: true, is_in_salon: false
```

as I working with sorted data I do reverse_each(it's my next PR 😉) to find overlaps and do splitting of them and one operation can produce up to 3 new items, to insert 3 elements in the middle of the array I have to resize it and offset tail of array, which is costly operation, with linked list it's easy one.

Here is how it looks in the end:

one of my backend test
```
      context do
        let(:working_hours) { { mon: { from: '8:00', to: '20:00' } } }
        let(:home_working_hours) { { mon: { from: '6:00', to: '24:00' } } }
        let!(:schedule_break) do
          create_break_for(
            salon,
            starts_at: date_of_next('mon').change(hour: 4, min: 0),
            ends_at: date_of_next('mon').change(hour: 14, min: 0)
          )
        end

        it do
          get "/vv/owner/salons/#{salon.id}/blocked_times",
              headers: auth_headers(owner),
              params: { date: Date.current }
          expect(blocked_times).to match_array(
            [
              { from: '00:00', to: '04:00', is_home: true, is_in_salon: true, reasons: ['Not working hours'] },
              { from: '04:00', to: '06:00', is_home: true, is_in_salon: true, reasons: ['Not working hours', "Schedule break '#{schedule_break.name}'"] },
              { from: '06:00', to: '08:00', is_home: true, is_in_salon: true, reasons: ['Not working hours', "Schedule break '#{schedule_break.name}'"] },
              { from: '08:00', to: '14:00', is_home: true, is_in_salon: true, reasons: ["Schedule break '#{schedule_break.name}'"] },
              { from: '20:00', to: '23:59', is_home: false, is_in_salon: true, reasons: ['Not working hours'] }
            ]
          )
        end
      end
```

and on front-end, I need to display this data like this:
 
![image](https://user-images.githubusercontent.com/1548210/44745500-91475e00-ab10-11e8-98e8-306847bc482d.png)

spectator replied:
That's very cool, man! Thanks for sharing!

spectator commented on lib/linked-list/list.rb:
```diff
@@ -191,19 +213,39 @@ def to_list
-      block = ->(e) { e == val } unless block_given?
-      -> (node) { block.call(node.data) }
+    def insert_before_node(data, node)
+      Node.new(data).tap do |new_node|
```
Let's use `Node(data)` similar to other methods.

mpospelov commented:
> 🤦‍♂️ I'm sorry, for some reason I thought first argument is "index", maybe because that is how Array#insert works. Maybe keep order of arguments similar to that?

We can, but that's not an index anyway :) What about keyword version?
```
insert(3, after: 1)
insert(3, after: (d) -> d == 1)
insert(3, before: 1)
insert(3, before: (d) -> d == 1)
```

spectator commented:
@mpospelov keyword version is good too!

mpospelov committed:
implement keyword version of insert

mpospelov committed:
make unlink private
